community/kernel/src/main/java/org/neo4j/unsafe/impl/batchimport/cache/DynamicNumberArray.java

```diff
@@ -28,69 +28,17 @@
  * @see NumberArrayFactory#newDynamicLongArray(long, long)
  * @see NumberArrayFactory#newDynamicIntArray(long, int)
  */
-abstract class DynamicNumberArray<N extends NumberArray> implements NumberArray
+abstract class DynamicNumberArray<N extends NumberArray> extends ChunkedNumberArray<N>
 {
     protected final NumberArrayFactory factory;
-    private final long chunkSize;
-    private NumberArray[] chunks = new NumberArray[0];
 
     DynamicNumberArray( NumberArrayFactory factory, long chunkSize )
     {
+        super( chunkSize );
         this.factory = factory;
-        this.chunkSize = chunkSize;
+        this.chunks = new NumberArray[0];
     }
 
-    @Override
-    public long length()
-    {
-        return chunks.length * chunkSize;
-    }
-
-    @Override
-    public long size()
-    {
-        long size = 0;
-        for ( int i = 0; i < chunks.length; i++ )
-        {
-            size += chunks[i].size();
-        }
-        return size;
-    }
-
-    @Override
-    public void clear()
-    {
-        for ( NumberArray chunk : chunks )
-        {
-            chunk.clear();
-        }
-    }
-
-    @Override
-    public long highestSetIndex()
-    {
-        for ( int i = chunks.length-1; i >= 0; i-- )
-        {
-            NumberArray chunk = chunks[i];
-            long highestSetInChunk = chunk.highestSetIndex();
-            if ( highestSetInChunk > -1 )
-            {
-                return i*chunkSize + highestSetInChunk;
-            }
-        }
-        return -1;
-    }
-
-    @Override
-    public void acceptMemoryStatsVisitor( MemoryStatsVisitor visitor )
-    {
-        for ( NumberArray chunk : chunks )
-        {
-            chunk.acceptMemoryStatsVisitor( visitor );
-        }
-    }
-
-    @SuppressWarnings( "unchecked" )
     protected N ensureChunkAt( long index )
     {
```

Comment:
Yes and somehow hotspot had trouble inlining things or in some way optimizing things when that was on the path of execution so having a version that doesn't use it made a huge impact.